osgp/protocol-adapter-mqtt/osgp-protocol-simulator-mqtt/src/main/java/org/opensmartgridplatform/simulator/protocol/mqtt/Client.java

```diff
@@ -27,30 +37,75 @@
+    this.mqttClientProperties = mqttClientProperties;
     this.uuid = UUID.randomUUID();
   }
 
   @Override
   public void run() {
     this.running = true;
-    this.mqtt3BlockingClient =
-        Mqtt3Client.builder()
-            .identifier(this.uuid.toString())
-            .serverHost(this.host)
-            .serverPort(this.port)
-            .buildBlocking();
-    final Mqtt3ConnAck ack = this.mqtt3BlockingClient.connect();
-    LOG.info(
-        String.format("Client %s received Ack %s", this.getClass().getSimpleName(), ack.getType()));
-    this.addShutdownHook();
-    LOG.info(String.format("Client %s started", this.getClass().getSimpleName()));
-    this.onConnect(this.mqtt3BlockingClient);
+
+    try {
+      this.mqtt3BlockingClient =
+          Mqtt3Client.builder()
+              .identifier(this.uuid.toString())
+              .serverHost(this.host)
+              .serverPort(this.port)
+              .sslConfig(getSslConfig(this.mqttClientProperties))
+              .buildBlocking();
+
+      final Mqtt3ConnAck ack = this.mqtt3BlockingClient.connect();
+      LOG.info("Client {} received Ack {}", this.getClass().getSimpleName(), ack.getType());
+      this.addShutdownHook();
+      LOG.info("Client {} started", this.getClass().getSimpleName());
+      this.onConnect(this.mqtt3BlockingClient);
+    } catch (final Exception ex) {
+      LOG.error("Exception while starting client.", ex);
+      this.stopClient();
+    }
+  }
+
+  private static MqttClientSslConfig getSslConfig(final Properties mqttClientProperties)
+      throws GeneralSecurityException {
+    if (mqttClientProperties == null || mqttClientProperties.isEmpty()) {
+      return null;
+    }
+    return MqttClientSslConfig.builder()
+        .trustManagerFactory(getTruststoreFactory(mqttClientProperties))
+        .build();
+  }
+
+  private static TrustManagerFactory getTruststoreFactory(final Properties mqttClientProperties)
+      throws GeneralSecurityException {
+
+    final String trustStoreType =
+        mqttClientProperties.getProperty(ClientConstants.SSL_TRUSTSTORE_TYPE_PROPERTY_NAME);
+    final String trustStorePath =
+        mqttClientProperties.getProperty(ClientConstants.SSL_TRUSTSTORE_PATH_PROPERTY_NAME);
+    final String trustStorePW =
+        mqttClientProperties.getProperty(ClientConstants.SSL_TRUSTSTORE_PASSWORD_PROPERTY_NAME);
+
+    KeyStore trustStore;
+    if (Files.exists(Paths.get(trustStorePath))) {
+      LOG.info("Load external truststore from path: {}", trustStorePath);
+      trustStore = loadTrustStoreFromExternalPath(trustStoreType, trustStorePath, trustStorePW);
+    } else {
+      LOG.info("Load default truststore from classpath: {}", trustStorePath);
+      trustStore = loadTrustStoreFromClassPath(trustStoreType, trustStorePath, trustStorePW);
+    }
+
+    final TrustManagerFactory tmf =
+        TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+    tmf.init(trustStore);
+
+    return tmf;
   }
```

Comment:
This looks almost identical to the code in `MqttClientAdapter` (lines 96 to 130) in osgp-protocol-adapter-mqtt.
Some form of more specific keystore configuration object, or maybe even MQTT specific could probably remove the duplication and simplify the code where it is used.
The code for alternatives for loading the keystore files (140-167) would fall under this same comment I'd guess.

Reply:
Refactored